### DISCUSSION TITLE
How to filter EBSD image ?

### DISCUSSION THREAD

**Discussion by Grammatica1312** [mtex-toolbox/mtex#535] [Ask Anything]
Hi everyone,

I'm Michela, a PhD Student at the University of Milan. I performed several experiments to study the melt-rock reaction.
One of the aim of my research is to study the textural evolution of olivine. For this reason, i performed EBSD analysis. 
I have already a script  adapted for different textural analysis like IPF, GOS, Misorientation and so on. However, i have some problems to filter my maps because i have a lot of non-indexed point. Can you help me to solve this problem? Unfortunately, i have no programming experience, so i help you can help me. 
I attached some example of unfiltered maps, hoping it will help 

Thanks for your help
Cheers,
Michela
![MG_1_4_map_Mean_Angular_Deviation_initial](https://user-images.githubusercontent.com/73173170/96583265-ba24a180-12dc-11eb-8d3f-ae99406710f9.png)

![MG_1_4_map_Olivine_orientation_ipfcolourbar_x](https://user-images.githubusercontent.com/73173170/96583295-c7da2700-12dc-11eb-96b5-e02df9c998da.png)

**Comment by ralfHielscher** [comment]
Hi Michela,

could you please tell problems you faced denoising your data with MTEX? Pages you should have read

* https://mtex-toolbox.github.io/EBSDDenoising.html
* https://mtex-toolbox.github.io/EBSDFilling.html
* https://mtex-toolbox.github.io/GrainReconstructionAdvanced.html

Ralf.

**Reply by Grammatica1312** [reply]
Hi Ralf,

thank you for answered. 
In my script, i have a mix of the three routine that you suggested me or probably some part of them. I tried to work with grain reconstruction routine (not advanced) but in this way i was losing the melt component. 
I attached a part of my script, in which there is some function useful to filter my map, in this is too easy to understand what i need to add or not to my script. 
```matlab
%% GRAIN DETECTION - DETECTING AND REMOVING TWINS - FILTER OUT SMALL GRAINS %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Grain modelling
% You can loop over this cell changing segmentation angle and the 'keepNotIndexed' options
% You should compared the resulting grain map with EBSD (pixel) phase or orientation (pixel) map or Band Contrast map
close all
disp(' ')
disp(' Grain segmentation angle option ')
disp(' Choose a high angle typically between 15 to 10 degrees for geological samples')
disp(' OR choose low angle of 2 degrees if you want to detect sub-grains')
%segmentation_angle = input('The segmentation angle (e.g. 2-15):');
segmentation_angle = 10
segAngle = segmentation_angle*degree;
%
disp(' ')
disp(' Keep non-indexed points option ')
disp('*1= Scientifically correct, not extrapolating raw indexed data')
disp('    model grains BUT keep non-index points : Default in MTEX 4')
disp(' 2= May be more geologically correct in some cases, use with care')
disp('    model grains AND include non-index points within grains boundaries')
disp('    N.B. this option does NOT ADD map pixels with neighbouring ORIENTATIONS')
%non_indexed_option = input('Option an integer  (1-2):');
non_indexed_option = 1
%
% keep non-indexed in is default in MTEX4
if(non_indexed_option == 1)
    [grains,ebsd_clean.grainId,ebsd_clean.mis2mean] = calcGrains(ebsd_clean,'boundary','tight','angle',segAngle)
end
% remove non-indexed points, restrict to indexed points only
if(non_indexed_option == 2)
    [grains,ebsd_clean.grainId,ebsd_clean.mis2mean] = calcGrains(ebsd_clean('indexed'),'boundary','tight','angle',segAngle)
end
% smooth grains
grains = smooth(grains,5)
% number of grains all phases 
number_of_grains = grains.length
number_of_plag_grains = grains('Plagioclase').length
number_of_cpx_grains = grains('Clinopyroxene').length
number_of_ol_grains = grains('Olivine').length
number_of_opx_grains = grains('Orthopyroxene').length
number_of_mag_grains = grains('Magnetite').length
number_of_ilm_grains = grains('Ilmenite').length
% write to output file
fprintf(output_file,'\n\n');
fprintf(output_file,'Segmentation angle for grain detection : %i \n',segmentation_angle)
fprintf(output_file,'Keep non-indexed points option : %i \n' ,non_indexed_option)
fprintf(output_file,'Initial number of calculated grains : %i \n',number_of_grains)
fprintf(output_file,'Initial number of calculated plagioclase grains : %i \n',number_of_plag_grains)
fprintf(output_file,'Initial number of calculated clinopyroxene grains : %i \n',number_of_cpx_grains)
fprintf(output_file,'Initial number of calculated olivine grains : %i \n',number_of_ol_grains)
fprintf(output_file,'Initial number of calculated Orthopyroxene grains : %i \n',number_of_opx_grains)
fprintf(output_file,'Initial number of calculated Magnetite grains : %i \n',number_of_mag_grains)
fprintf(output_file,'Initial number of calculated Ilmenite grains : %i',number_of_ilm_grains)
% plot 'grain' phase map
%close all
% Default phase map
% plot(grains('indexed'),'figSize','large')

%% remove small grains - not representative small grains, may be errors
% You can loop over this cell changing number of indexed points per grain
% You should compared the resulting grain map with EBSD (pixel) phase or orientation (pixel) map
disp(' ')
disp(' Small grains option ')
disp(' Remove small grains containing less than a critical')
disp(' number of indexed points as they error prone or ')
disp(' If you require an accurate grain size and shape analysis')
disp(' the recommended minimum number indexed points per grain size is 10')
disp(' You can decide to keep all grain by accepting all grains with 0')
%small_grains_option = input('Indexed points per grain an integer (e.g. 0-10):');
small_grains_option = 0
% remove grains containing less than critical number of indexed points, 
grains_selected = grains(grains.grainSize > small_grains_option);
%grains_selected = grains(grains.grainSize > 10 & grains.grainSize < 50);
% Definition of ebsd for selected grains
ebsd_selected = ebsd_clean(grains_selected)
ebsd_smoothed = smooth(ebsd_selected('indexed'),splineFilter,'fill')
ebsd_smoothed = smooth(ebsd_selected('indexed'),splineFilter,'fill',grains)
```
I attached you other example of maps obtained applying the SMOOTH, how you can see the grain are partially fill but there are also a lot of point or areas not indexed.
![MG_10_map_Olivine_Orientation_GrainBoundaries](https://user-images.githubusercontent.com/73173170/96588499-03c4ba80-12e4-11eb-93c4-9ed121991bbc.png)
![MG_10_map_Olivine_Mis2Mean_GrainBoundaries](https://user-images.githubusercontent.com/73173170/96588530-0cb58c00-12e4-11eb-8613-86def99792ab.png)
![MG_10_map_OlivineGrains_GOS](https://user-images.githubusercontent.com/73173170/96588566-19d27b00-12e4-11eb-9605-3fb1f12896fe.png)

Thank you in advance

**Comment by ralfHielscher** [comment]
Hi Michaela,

I still do not understand what the problem is. Would you like to fill the small grains?

Ralf.

**Reply by Grammatica1312** [reply]
Yes i would fill that grains but the other problem is that in the maps there are a lot of white area. In these areas would be some grains not only glass and the problem is that i don't know how to reconstruct that grains that are not visible in the maps

Michela

**Reply by kilir** [reply]
How would you decide which part of the non-indexed areas (in your initial map) belongs to a grain or to the glass?

Cheers,
Rüdiger

**Reply by Grammatica1312** [reply]
Hi,

i previously analyzed this sample with microprobe that provides chemical analysis and BSE images. Based on this i compared EBSD images with BSE images and i can distinguish glass (melt) from grains.

Michela

**Comment by zmichels** [comment]
Hi Michela,

This is a common problem with collecting EBSD on melt-bearing samples and I know how frustrating it can be, especially for recovering any tiny grains in a melt matrix. I believe the reason it is common is because it is very difficult to identically and sufficiently polish the melt *and* the tiny grains within it, such that EBSD acquisition is successful and you are able to index the tiny/suspended grains. Without an orientation solution, you cannot reconstruct grains from the EBSD data. You could try to analyze the image/map of the band contrast using other tools in Matlab perhaps, but there is not currently an MTEX routine for defining grains from band contrast maps. 

Are the data “cleaned” at all (i.e., any points or “wild spikes” thrown out) prior to importing in MTEX? If so, you might want to ry with an uncleaned version of the data and keep any points in the white areas.

Cheers,
Zach

**Reply by kilir** [reply]
Hi,
one can actually do something like this for example:
```
mtexdata ferrite
grains = ebsd.calcGrains(ebsd.ci,0.2)
```
Cheers,
Rüdiger

**Reply by Grammatica1312** [reply]
Hi Zach,

I also tried to analyze map of band contrast but unfortunately is not good, i attached you two images: the first one represents the map of initial band contrast and the second one represents the band contrast of selected phases after the elaboration with MTEX .
The data are not "cleaned" at all prior to importing in MTEX for several reasons. It's a bit tricky to get good maps or  good cleaning prior to import in MTEX. For this reason, i tried and i'm trying to find a solution to solve this problem.


![MG_10_map_Band_Contrast_initial](https://user-images.githubusercontent.com/73173170/96697208-ef3bfd00-138b-11eb-925a-8be293ca4407.png)
![MG_10_map_Band_Contrast_PhaseGrains](https://user-images.githubusercontent.com/73173170/96697224-f6fba180-138b-11eb-968a-51070d2126d3.png)

Michela

**Comment by zmichels** [comment]
Fantastic tip — thank you!
